File: clients/browser_javascript/src/simple_uploader.ts
import axios, { AxiosRequestConfig } from "axios";
import camecaseKeys from "camelcase-keys";
import { PromisePool } from '@supercharge/promise-pool'

interface Progress {
  allSlice: number;
  finishedSlice: number;
}


interface Response<T> {
  code: number;
  message: string;
  data: T;
}

interface Slice {
  sliceId: string;
  status: number;
  sha1: string;
}

interface FileMeta {
  fileId: string;
  fileName: string;
  fileType: string;
  fileSize: number;
  chunkSize: number;
  prefix: string;
  createdAt: number;
  status: number;
  slices: { [key: string]: Slice };
}

export class UserCanceledUploading extends Error {
  constructor() {
    super("User Canceled Upload");
    this.name = "UserCanceledUploading"
  }
}

function sha1File(file: Blob): Promise<string> {
  return new Promise((resolve, _) => {
    const reader = new FileReader();
    reader.onload = async () => {
      const res = await crypto.subtle.digest(
        "SHA-1",
        reader.result as ArrayBuffer
      );
      const hashArray = Array.from(new Uint8Array(res));
      const hashHex = hashArray
        .map((b) => b.toString(16).padStart(2, "0"))
        .join("");
      return resolve(hashHex);
    };
    reader.readAsArrayBuffer(file);
  });
}

type SimpleUploaderOptions = {
  chunkSize: number;
  endpoint: string;
  onProgress?: (progress: Progress) => void;
  onCheckingSumProgress?: (progress: Progress) => void;
  requestOptions?: AxiosRequestConfig;
  prefix: string;
  concurrent: number;
  useV2?: boolean;
};

export default class SimpleUploader {
  meta?: FileMeta | null;
  metaKey: string;
  public options: SimpleUploaderOptions;
  halt: boolean;

  constructor(public file: File, options: Partial<SimpleUploaderOptions>) {
    this.options = Object.assign(
      {
        concurrent: 4,
        chunkSize: 1024 * 1024 * 10,
        endpoint: "/files",
        onProgress: undefined,
        prefix: "",
        useV2: true,
      },
      options
    );
    this.meta = null;
    this.metaKey = `file_meta_${this.file.name}_${this.file.size}`;
    this.loadMeta();
    this.halt = false;
    if (this.meta) {
      console.log("meta data loaded from local storage", this.meta);
    }
  }

  clearMeta() {
    this.meta = null;
    localStorage.removeItem(this.metaKey);
  }

  loadMeta() {
    const data = localStorage.getItem(this.metaKey);
    if (data) {
      try {
        this.meta = JSON.parse(data);
      } catch {
        console.log(`failed to parse meta data for file ${this.file.name}`);
      }
    }
  }

  cancel() {
    this.halt = true;
  }

  async upload() {
    if (!this.meta) {
      const response = await axios.post<Response<FileMeta>>(
        this.options.endpoint,
        {
          file_name: this.file.name,
          file_type: this.file.type,
          file_size: this.file.size,
          chunk_size: this.options.chunkSize,
          prefix: this.options.prefix
        },
        this.options.requestOptions
      );
      if (response.status !== 200) {
        throw new Error(response.statusText);
      }
      this.meta = camecaseKeys(response.data.data, { deep: true });
      this.saveMeta();
    }

    const slicesIds = Object.keys(this.meta!.slices)

    let error: Error | null = null
    await PromisePool
      .for(slicesIds)
      .withConcurrency(this.options.concurrent)
      .handleError(async (e, _user, pool) => {
        if (e) {
          error = e
        }
        pool.stop();
      })
      .process(async (sliceId) => {
        if (this.halt) {
          this.halt = false;
          throw new UserCanceledUploading();
        }
        const slice = this.meta!.slices[sliceId];
        if (slice.status === 0) {
          const response = await this.uploadSlice(slice.sliceId);
          if (response.code == 206 || response.code == 200) {
            this.meta!.slices[slice.sliceId].status = 1;
            this.saveMeta();
          }
          if (this.options.onProgress) {
            this.options.onProgress({
              allSlice: Object.keys(this.meta!.slices).length,
              finishedSlice: Object.values(this.meta!.slices).filter(
                (s) => s.status === 1
              ).length,
            });
          }
        }
      })

    if (error) {
      console.log("throwing error: ", error)
      throw error
    }
  }

  async uploadSlice(sliceId: string): Promise<Response<string>> {
    const formData = new FormData();
    const sliceIdInt = parseInt(sliceId);
    formData.append(
      "file",
      this.file.slice(
        sliceIdInt * this.options.chunkSize,
        (sliceIdInt + 1) * this.options.chunkSize
      ),
      this.file.name
    );
    formData.append("slice_id", sliceId);
    formData.append("file_id", this.meta!.fileId);
    formData.append("file_name", this.meta!.fileName);
    formData.append("file_type", this.meta!.fileType);
    formData.append("file_size", this.meta!.fileSize.toString());
    formData.append("chunk_size", this.options.chunkSize.toString());

    const response = await axios.post<Response<string>>(
      `${this.options.endpoint}/${this.meta!.fileId}/upload${this.options.useV2 ? '_v2' : ''}`,
      formData,
      {
        ...this.options.requestOptions,
        headers: {
          ...(this.options.requestOptions &&
            this.options.requestOptions.headers),
          "Content-Type": "multipart/form-data",
        },
      }
    );

    if (response.status >= 400) {
      throw new Error(
        `failed to upload slice ${sliceId}, error: ${response.statusText}`
      );
    }

    return response.data;
  }

  saveMeta() {
    localStorage.setItem(this.metaKey, JSON.stringify(this.meta));
  }

  async checksum(): Promise<void> {
    const serverMeta = await axios.get<Response<FileMeta>>(
      `${this.options.endpoint}/${this.meta!.fileId}/meta`,
      this.options.requestOptions
    );
    let checksumIndex = 1;
    for (let sliceId in this.meta!.slices) {
      const slice = this.meta!.slices[sliceId];
      const sliceIdInt = parseInt(slice.sliceId);
      const file = this.file.slice(
        sliceIdInt * this.options.chunkSize,
        (sliceIdInt + 1) * this.options.chunkSize
      );
      const sha1 = await sha1File(file);

      if (serverMeta.data.data.slices[sliceId].sha1 !== sha1) {
        console.log(`slice ${sliceId} sha1 mismatch, reupload the slice`)
        let response = await this.uploadSlice(sliceId);
        if (response.code == 206 || response.code == 200) {
          console.log(`slice ${sliceId} reupload success`)
        } else {
          console.log(`slice ${sliceId} reupload failed`)
          // set the slice as not uploaded and then raise error
          this.meta!.slices[sliceId].status = 0;
          this.saveMeta();
          throw new Error("checksum failed")
        }
      }

      if (this.options.onCheckingSumProgress) {
        this.options.onCheckingSumProgress({
          allSlice: Object.keys(this.meta!.slices).length,
          finishedSlice: checksumIndex,
        });
        checksumIndex += 1;
      }
    }
  }
}

File: clients/python/simple_uploader/simple_uploader.py
import hashlib
import io
import json
import os
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any, Callable, Dict, List, Optional, Union

import requests
from dacite import from_dict


@dataclass
class CheckResult:
    success_count: int
    failed_count: int
    failed_slices_id: List[str]


@dataclass
class Progress:
    all_slice: int
    finished_slice: int


@dataclass
class Slice:
    slice_id: str
    status: int
    sha1: str


@dataclass
class FileMeta:
    file_id: str
    file_name: str
    file_type: str
    file_size: int
    chunk_size: int
    prefix: str
    created_at: int
    status: int
    slices: Dict[str, Slice]


@dataclass
class Options:
    chunk_size: Optional[int] = 1024 * 1024 * 10
    endpoint: Optional[str] = "http://127.0.0.1:8080/files"
    on_progress: Optional[Callable[[Progress], None]] = None
    prefix: Optional[str] = ""
    headers: Optional[Dict[str, str]] = None
    use_v2: Optional[bool] = True


@dataclass
class Response:
    code: int
    message: str
    data: Optional[Any] = None


class SimpleUploader:
    file: Path
    options: Options
    meta: FileMeta
    meta_key: str
    meta_file: Path

    def __init__(
        self, file: Union[Path, str], options: Optional[Union[dict, Options]] = {}
    ):
        self.file = Path(file)
        self.fh = self.file.open("rb")
        self.file_size = os.path.getsize(self.file)
        if isinstance(options, dict):
            self.options = Options(**options)
        self.meta = None
        self.meta_key = f"file_meta_{self.file.name}_{self.file_size}"
        self.meta_file = (
            Path.home() / ".cache" / "simple_uploader" / (self.meta_key + ".json")
        )
        self.meta_file.parent.mkdir(parents=True, exist_ok=True)
        self.meta_file.touch(exist_ok=True)
        self.load_meta()

    def clear_meta(self):
        self.meta = None
        self.meta_file.unlink()

    def load_meta(self):
        try:
            self.meta = from_dict(FileMeta, json.loads(self.meta_file.read_text()))
        except json.decoder.JSONDecodeError:
            pass

    def save_meta(self):
        self.meta_file.write_text(json.dumps(asdict(self.meta)))

    def upload(self):
        if not self.meta:
            response = requests.post(
                self.options.endpoint,
                json={
                    "file_name": self.file.name,
                    "file_type": self.file.suffix,
                    "file_size": self.file_size,
                    "chunk_size": self.options.chunk_size,
                    "prefix": self.options.prefix,
                },
                headers={
                    **(self.options.headers if self.options.headers is not None else {})
                },
            )
            if response.status_code != 200:
                raise Exception(response.text)
            self.meta = from_dict(FileMeta, response.json()["data"])
            self.save_meta()

        for slice_id, slice in self.meta.slices.items():
            if slice.status == 0:
                response = self._upload_slice(slice_id)
                if response.code == 206 or response.code == 200:
                    self.meta.slices[slice_id].status = 1
                    self.save_meta()
                if self.options.on_progress:
                    self.options.on_progress(
                        {
                            "all_slice": len(self.meta.slices),
                            "finished_slice": len(
                                list(
                                    filter(
                                        lambda s: s.status == 1,
                                        self.meta.slices.values(),
                                    )
                                )
                            ),
                        }
                    )

    def _upload_slice(self, slice_id: str) -> Response:
        self.fh.seek(int(slice_id) * self.options.chunk_size)
        bytes = self.fh.read(self.options.chunk_size)
        form_data = {
            "slice_id": slice_id,
            "file_id": self.meta.file_id,
            "file_name": self.meta.file_name,
            "file_type": self.meta.file_type,
            "file_size": self.meta.file_size,
            "chunk_size": self.options.chunk_size,
        }
        response = requests.post(
            f"{self.options.endpoint}/{self.meta.file_id}/upload{'_v2' if self.options.use_v2 else ''}",
            data=form_data,
            files={
                "file": io.BytesIO(bytes),
            },
            headers={
                **(self.options.headers if self.options.headers is not None else {})
            },
        )

        if response.status_code >= 400:
            raise Exception(response.text)

        return from_dict(Response, response.json())

    def _sha1(self, blob: bytes) -> str:
        sha1 = hashlib.sha1()
        sha1.update(blob)
        return sha1.hexdigest()

    def checksum(self) -> CheckResult:
        response = requests.get(
            f"{self.options.endpoint}/{self.meta.file_id}/meta",
            headers={
                **(self.options.headers if self.options.headers is not None else {})
            },
        )
        server_meta = from_dict(FileMeta, response.json()["data"])
        check_result = CheckResult(0, 0, [])
        for slice_id, slice in server_meta.slices.items():
            self.fh.seek(int(slice_id) * self.options.chunk_size)
            blob = self.fh.read(self.options.chunk_size)
            sha1 = self._sha1(blob)
            if slice.sha1 == sha1:
                check_result.success_count += 1
            else:
                check_result.failed_count += 1
                check_result.failed_slices_id.append(slice_id)
        return check_result
